system/HTTP/Request.php

```diff
@@ -92,7 +86,17 @@ public function isValidIP(?string $ip = null, ?string $which = null): bool
      */
     public function getMethod(bool $upper = false): string
     {
-        return ($upper) ? strtoupper($this->method) : strtolower($this->method);
+        if (empty($this->method)) {
+            $this->method = $this->getServer('REQUEST_METHOD') ?? 'GET';
+
+            if ($this->method === 'POST'
+                && in_array(strtoupper($_POST['_method'] ?? ''), ['PUT', 'PATCH', 'DELETE'], true)
```

Comment:
I'm not sure we need to limit the method here. From what I can find custom methods are allowed via the HTTP 1.1 specification, though not really encouraged. WebDav, for instance, has their own custom methods, I believe. 

If we are to limit, we need to include all of those that are in the spec, including HEAD, OPTIONS, TRACE, and CONNECT.  I'm interested to hear what others on the team have to say.

Reply:
AFAIK the HEAD, OPTIONS, TRACE, and CONNECT methods don't send any data in the message body. That's why I didn't include them in the test.

Reply:
Correct - but POST requests can have empty bodies so I see no reason not to support all of the standard methods. Otherwise we're needlessly limiting the available options.

Reply:
Right, you can still make a "POST OPTIONS" submission without content, and we need to be able to support that. I think leaving this open is the way to go. Nobody is going to submit a `_method` field without having some idea of what it does.

Reply:
The check I suggested is based on the documentation for the spoofing method, where PUT and DELETE methods are given as an example.
In any case, checking the spoofing method should be.

Reply:
I'm not an HTTP expert but at the end of the day if we weren't enforcing a method filter on spoofing previously then we shouldn't add it now. If there are legitimate uses of custom methods out there then this would break them. If you have compelling evidence for this to be a bug or a security issue then please submit it; otherwise please remove this for now, optionally submitting an issue to revisit for version 5.

Reply:
All user (may be an attacker) input should be validated.

Reply:
Validated or contained. In this case we may want to add a string regex filter `[A-Za-z]` or something more, but as long as the only place this is going is into the Request method I don't think there is a problem with leaving it up to user input. If someone wants to use POSTMAN to probe alternate route methods that's no different than some developer wanting to write an API based on methods named after fruits or Malazan characters.

Reply:
The framework supports routes based on HTTP methods (most). Until the main class for registering routes is extended, there can be no talk of any fruit methods.
POSTMAN supports custom HTTP methods.
The spoofing method was invented in order to send data at the POST request level for those HTTP methods that HTML form and PHP do not support.
You will not be able to submit form data via a PUT/PATCH/DELETE request. And PHP can't get that kind of data natively, only $_GET or $_POST.
This, and this alone, is the point of the spoofing method.

Can you give an example of why you need to send HEAD, OPTIONS, (TRACE, CONNECT - not supported by the framework) via an HTML form.

By the way, CSRF check is also tied to HTTP methods.

Reply:
I think there is no use case that sends requests other than methods `PUT/PATCH/DELETE` via HTML Form (method spoofing). The current behavior is a bug, and should be fixed.

Reply:
> If the form's method is not GET or POST, but PUT, PATCH or DELETE, Symfony will insert a hidden field with the name _method that stores this method. The form will be submitted in a normal POST request, but [Symfony's routing](https://symfony.com/doc/6.1/routing.html) is capable of detecting the _method parameter and will interpret it as a PUT, PATCH or DELETE request.
https://symfony.com/doc/6.1/forms.html#changing-the-action-and-http-method

> HTML forms do not support PUT, PATCH, or DELETE actions. So, when defining PUT, PATCH, or DELETE routes that are called from an HTML form, you will need to add a hidden _method field to the form. The value sent with the _method field will be used as the HTTP request method:
https://laravel.com/docs/8.x/routing#form-method-spoofing

> 'put', 'delete', 'patch' - Will override the HTTP method with PUT, DELETE or PATCH respectively, when the form is submitted.
https://book.cakephp.org/4/en/views/helpers/form.html#options-for-form-creation

Reply:
> as long as the only place this is going is into the Request method I don't think there is a problem with leaving it up to user input. 

It is used outside of the Request object.

> If someone wants to use POSTMAN to probe alternate route methods that's no different than some developer wanting to write an API based on methods named after fruits or Malazan characters.

CI4 Router does not support custom HTTP methods.
Auto-routing just ignores or does not consider HTTP methods.

Reply:
I'll wait